ipn/ipnlocal/local.go

```diff
@@ -2064,6 +2084,40 @@ func (b *LocalBackend) updateFilterLocked(netMap *netmap.NetworkMap, prefs ipn.P
 	}
 }
 
+// captivePortalWarnable is a Warnable which is set to an unhealthy state when a captive portal is detected.
+var captivePortalWarnable = health.Register(&health.Warnable{
+	Code:  "captive-portal-detected",
+	Title: "Captive portal detected",
+	// High severity, because captive portals block all traffic and require user intervention.
+	Severity:            health.SeverityHigh,
+	Text:                health.StaticMessage("This Wi-Fi network requires you to log in manually."),
+	ImpactsConnectivity: true,
+})
+
+// performCaptiveDetection checks if captive portal detection is enabled via controlknob. If so, it runs
+// the detection and updates the Warnable accordingly.
+func (b *LocalBackend) performCaptiveDetection() {
+	captiveDetectionDisabledByControlKnob := b.ControlKnobs().DisableCaptivePortalDetection.Load()
+	if captiveDetectionDisabledByControlKnob {
+		b.logf("performCaptiveDetection: disabled by controlknob")
+		return
+	}
+
+	d := captivedetection.NewDetector()
+	var dm *tailcfg.DERPMap
+	b.mu.Lock()
+	if b.netMap != nil {
+		dm = b.netMap.DERPMap
+	}
+	b.mu.Unlock()
+	found := d.Detect(b.ctx, b.NetMon(), dm, 0, b.logf)
+	if found {
```

Comment:
so "found" means the network was up enough for us to be 100% confident it's a captive portal, and false means no captive portal or unknown because network's not up enough to tell?

Reply:
Yup, that's the idea. It's a best-effort service. It only returns true if we're quite sure there is one, false in any other case.